src/tiled/stampbrush.cpp

```diff
@@ -386,6 +401,52 @@ void StampBrush::drawPreviewLayer(const QVector<QPoint> &list)
             mapDocument()->unifyTilesets(variation.map, mMissingTilesets);
 
             TileLayer *stamp = variation.tileLayer();
+            bool copyFlag = false;
+
+            //if staggered map, makes sure stamp stays the same
+            if ((mapDocument()->map()->orientation() == Map::Hexagonal
+                    || mapDocument()->map()->orientation() == Map::Staggered)
+                    && ((mapDocument()->map()->staggerAxis() == Map::StaggerY)?
+                         stamp->height() > 1 : stamp->width() > 1)) {
+                stamp = variation.tileLayer()->copy(variation.tileLayer()->bounds());
+                copyFlag = true;
```

Comment:
Oops, I totally forgot about the randomVariation part. You could still do a cache like:
```c++
QHash<TileLayer*, TileLayer*> shiftedCopies;

...

qDeleteAll(shiftedCopies);
```